provider/gandi/gandi_test.go

```diff
@@ -0,0 +1,755 @@
+		{
+			RrsetType:   endpoint.RecordTypeCNAME,
+			RrsetTTL:    600,
+			RrsetName:   "@",
+			RrsetHref:   "https://api.gandi.net/v5/domain/domains/example.com/records/%40/A",
+			RrsetValues: []string{"192.168.0.1"},
+		},
+		{
+			RrsetType:   endpoint.RecordTypeCNAME,
+			RrsetTTL:    600,
+			RrsetName:   "www",
+			RrsetHref:   "https://api.gandi.net/v5/domain/domains/example.com/records/www/CNAME",
+			RrsetValues: []string{"lb.example.com"},
+		},
+		{
+			RrsetType:   endpoint.RecordTypeA,
+			RrsetTTL:    600,
+			RrsetName:   "test",
+			RrsetHref:   "https://api.gandi.net/v5/domain/domains/example.com/records/test/A",
+			RrsetValues: []string{"192.168.0.2"},
+		},
+	}
+}
+
+// Mock all methods
+
+func (m *mockGandiClient) GetDomainRecords(fqdn string) (records []livedns.DomainRecord, err error) {
+	m.Actions = append(m.Actions, MockAction{
+		Name: "GetDomainRecords",
+		FQDN: fqdn,
+	})
+	if m.FunctionToFail == "GetDomainRecords" {
+		return nil, fmt.Errorf("injected error")
+	}
+
+	return m.RecordsToReturn, err
+}
+
+func (m *mockGandiClient) CreateDomainRecord(fqdn, name, recordtype string, ttl int, values []string) (response StandardResponse, err error) {
+	m.Actions = append(m.Actions, MockAction{
+		Name: "CreateDomainRecord",
+		FQDN: fqdn,
+		Record: livedns.DomainRecord{
+			RrsetType:   recordtype,
+			RrsetTTL:    ttl,
+			RrsetName:   name,
+			RrsetValues: values,
+		},
+	})
+	if m.FunctionToFail == "CreateDomainRecord" {
+		return StandardResponse{}, fmt.Errorf("injected error")
+	}
+
+	return StandardResponse{}, nil
+}
+
+func (m *mockGandiClient) DeleteDomainRecord(fqdn, name, recordtype string) (err error) {
+	m.Actions = append(m.Actions, MockAction{
+		Name: "DeleteDomainRecord",
+		FQDN: fqdn,
+		Record: livedns.DomainRecord{
+			RrsetType: recordtype,
+			RrsetName: name,
+		},
+	})
+
+	if m.FunctionToFail == "DeleteDomainRecord" {
+		return fmt.Errorf("injected error")
+	}
+
+	return nil
+}
+
+func (m *mockGandiClient) UpdateDomainRecordByNameAndType(fqdn, name, recordtype string, ttl int, values []string) (response StandardResponse, err error) {
+	m.Actions = append(m.Actions, MockAction{
+		Name: "UpdateDomainRecordByNameAndType",
+		FQDN: fqdn,
+		Record: livedns.DomainRecord{
+			RrsetType:   recordtype,
+			RrsetTTL:    ttl,
+			RrsetName:   name,
+			RrsetValues: values,
+		},
+	})
+
+	if m.FunctionToFail == "UpdateDomainRecordByNameAndType" {
+		return StandardResponse{}, fmt.Errorf("injected error")
+	}
+
+	return StandardResponse{}, nil
+}
+
+func (m *mockGandiClient) ListDomains() (domains []domain.ListResponse, err error) {
+	m.Actions = append(m.Actions, MockAction{
+		Name: "ListDomains",
+	})
+
+	if m.FunctionToFail == "ListDomains" {
+		return []domain.ListResponse{}, fmt.Errorf("injected error")
+	}
+
+	return []domain.ListResponse{
+		{
+			FQDN:        "example.com",
+			FQDNUnicode: "example.com",
+			Href:        "https://api.gandi.net/v5/domain/domains/example.com",
```

Comment:
I'd suggest adding this as a const or something, so in case the version gets updates, you don't need to update on each line 🙂